### ISSUE TITLE
IIIF - possible wrong calculation of tile width

### ISSUE THREAD

**Issue by filak** [openseadragon/openseadragon#2198]
I have this image served by IIPImage server:

```
{
"@context": "http://iiif.io/api/image/2/context.json",
"protocol": "http://iiif.io/api/image",
"width": 2609,
"height": 3424,
"sizes": [
{
"width": 163,
"height": 214
},
{
"width": 326,
"height": 428
},
{
"width": 652,
"height": 856
},
{
"width": 1304,
"height": 1712
}
],
"tiles": [
{
"width": 256,
"height": 256,
"scaleFactors": [
1,
2,
4,
8,
16
]
}
],
"@id": "http://iip.test/iiif/f.jp2",
"profile": [
"http://iiif.io/api/image/2/level2.json",
{
"formats": [
"jpg",
"png",
"webp"
],
"qualities": [
"native",
"color",
"gray",
"bitonal"
],
"supports": [
"regionByPct",
"regionSquare",
"sizeByForcedWh",
"sizeByWh",
"sizeAboveFull",
"sizeUpscaling",
"rotationBy90s",
"mirroring"
],
"maxWidth": 3500,
"maxHeight": 3500
}
]
}
```

When used with Openseadragon viewer I get these IIIF API calls:

First row:
- http://iip.test/iiif/f.jp2/0,0,2048,2048/256,/0/default.jpg => 256 x 256
- http://iip.test/iiif/f.jp2/2048,0,561,2048/71,/0/default.jpg => **71** x **259**

Second row:
- http://iip.test/iiif/f.jp2/0,2048,2048,1376/256,/0/default.jpg => 256 x 172
- http://iip.test/iiif/f.jp2/2048,2048,561,1376/71,/0/default.jpg => **71** x **174**

The content of the last tile column appears dropped a bit when zoomig out or on a small displays.

![iip_iiif_osd](https://user-images.githubusercontent.com/1353748/186459022-7c246bd2-df4e-402b-8116-30054c5eac76.png)

As @ruven kindly pointed out at https://github.com/ruven/iipsrv/issues/237#issuecomment-1235444340 it looks like the width parameter seems probably miscalculated resulting in bad tile height.

**Comment by iangilman** [comment]
Interesting! Thank you for filing this, and for linking to the additional information. I'm not terribly familiar with the IIIF tile source, so I don't know if this is specific to that tile source or if it's something more general in our code. Can you help me understand what's going on? Is OpenSeadragon asking for a specific height? If so, how is it doing that? I'm seeing in your example this URL: 

http://iip.test/iiif/f.jp2/2048,2048,561,1376/71,/0/default.jpg

... And I'm seeing 71 in it, but not 174. Do you know where the 174 is coming from? 

BTW, if you'd be interested in taking a closer look at the IIIF code, this is the file: 

https://github.com/openseadragon/openseadragon/blob/master/src/iiiftilesource.js

@azaroth42 Any thoughts on this issue? BTW, thank you for letting me continue to tag you with IIIF issues. Let me know if/when someone else would be more appropriate :)

**Comment by azaroth42** [comment]
Could you try modifying your local copy for these lines:
    https://github.com/openseadragon/openseadragon/blob/master/src/iiiftilesource.js#L401-L402 
to use floor instead of ceil, and see if it works as expected?

`(2609 - 2048) / 8` is 70.125, and the ceil makes it a width of 71. The top left of the region is correct in the second column, as is the width (561) and height (2048). So the only other parameter to manipulate is the desired size which has to be a whole number of pixels. Asking for a width of 71, when the number of pixels is 561 is 0.12656, times the height of 2048 is 259, giving you the three undesirable pixels.

That said, this has been working for a long time now so I'm surprised it hasn't come up before

**Comment by iangilman** [comment]
@azaroth42 Thank you! If this does make the difference, I wonder if making the change would break other things? Also, what about `round` instead?

**Comment by ruven** [comment]
> Also, what about round instead?

The IIPImage server [uses round() internally](https://github.com/ruven/iipsrv/blob/ffe319016e6b7de57778083a033b7b320e3a9247/src/IIIF.cc#L448) to calculate these sizes for IIIF.

**Comment by iangilman** [comment]
@ruven Good to know! So round might actually be more correct for us to use. @filak can you confirm it fixes your issue?

**Comment by filak** [comment]
I am not sure how to build your code so I cannot confirm.

**Comment by filak** [comment]
Would you be so kind and build a dev release so I can test it?

**Comment by iangilman** [comment]
@filak I generally don't do a custom builds for people. In this case, if it works, the next step would be to make a pull request. If you'd already made the change yourself, you'd be one step away from submitting a patch and making it an official fix.

Here are the build instructions:

https://github.com/openseadragon/openseadragon/blob/master/CONTRIBUTING.md

If this is a problem for you, though, I can see about doing a custom build when I get a chance next week.

**Comment by filak** [comment]
OK, I have created a PR https://github.com/openseadragon/openseadragon/pull/2206

**Comment by filak** [comment]
@iangilman Would you be so kind and review the PR ?

**Comment by iangilman** [comment]
@filak Awesome... Thank you for making the PR! Apologies for the delay... Last week was a little kooky. :)

**Comment by iangilman** [comment]
Fixed by #2206.
